### ISSUE TITLE
Incorrect resolution of Gradle configuration dependencies in submodules

### ISSUE THREAD

**Issue by cusxy** [gradle/gradle#26767]
### Current Behavior

- The configuration of the root Gradle module depends on the library org.example:library:1.0.0
- The configuration of the Gradle sub-module depends on the library org.example:library:1.1.0.

When the configuration is resolved, the org.example:library:1.0.0 version is used.

### Expected Behavior

When the configuration is resolved, the org.example:library:1.1.0 version is used.

### Context (optional)

When updating the versions of the kotlinx-serialization library in our convention plugins, we got a NoSuchMethodException error in submodules in almost all of our composite builds.

To solve the problem, we now explicitly duplicate the versions of the libraries used to customize the build configuration in convention plugins in the dependencies block of the buildscript.

This error is reproduced with any libraries.

### Steps to Reproduce

[Here is an example project](https://github.com/cusxy/gradle-dependencies-project) where you can reproduce the error when using gradle convention plugin and regular gradle plugin.

To reproduce the error you can uncomment any line where plugins are plugged in the file [sample/app/build.grade.kts](https://github.com/cusxy/gradle-dependencies-project/blob/master/sample/app/build.gradle.kts).

### Gradle version

8.4

**Comment by ov7a** [comment]
Thank you for your interest in Gradle!

This is the implemented behavior, and there’s no plan to change it. This issue will be closed.

***

The root cause for this is the classloader hierarchy. The root project classloader has priority. To solve this, you can remove the plugin from the root project, move anything that needs this plugin into a separate subproject, and apply the plugin with a different version.

In short, to apply different versions of plugins, they can not have a parent-child relationship.
